<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.56">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Jan Emmanuel Samson">
<meta name="dcterms.date" content="2024-08-28">
<meta name="description" content="Build trees to infer phylogenetic relationships across species.">

<title>Modeling Evolutionary Relationships – BioHub</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">BioHub</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../notebook.html"> 
<span class="menu-text">Notebook</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../concepts.html"> 
<span class="menu-text">Concepts</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../algorithms.html"> 
<span class="menu-text">Algorithms</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../rosalind.html"> 
<span class="menu-text">Rosalind</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../about.html"> 
<span class="menu-text">Author</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
    <a href="https://x.com/dagsdags13" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/dagsdags212" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-github"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#evolution" id="toc-evolution" class="nav-link active" data-scroll-target="#evolution">Evolution</a>
  <ul class="collapse">
  <li><a href="#what-drives-change" id="toc-what-drives-change" class="nav-link" data-scroll-target="#what-drives-change">What drives change?</a></li>
  <li><a href="#how-do-we-observe-selection" id="toc-how-do-we-observe-selection" class="nav-link" data-scroll-target="#how-do-we-observe-selection">How do we observe selection?</a></li>
  <li><a href="#synonymous-vs.-non-synonymous-mutations" id="toc-synonymous-vs.-non-synonymous-mutations" class="nav-link" data-scroll-target="#synonymous-vs.-non-synonymous-mutations">Synonymous vs.&nbsp;non-synonymous mutations</a></li>
  </ul></li>
  <li><a href="#phylogenetics" id="toc-phylogenetics" class="nav-link" data-scroll-target="#phylogenetics">Phylogenetics</a>
  <ul class="collapse">
  <li><a href="#rooted-and-unrooted-trees" id="toc-rooted-and-unrooted-trees" class="nav-link" data-scroll-target="#rooted-and-unrooted-trees">Rooted and unrooted trees</a></li>
  <li><a href="#how-do-we-generate-a-tree" id="toc-how-do-we-generate-a-tree" class="nav-link" data-scroll-target="#how-do-we-generate-a-tree">How do we generate a tree?</a></li>
  <li><a href="#substitution-models" id="toc-substitution-models" class="nav-link" data-scroll-target="#substitution-models">Substitution Models</a></li>
  </ul></li>
  <li><a href="#distance-based-methods" id="toc-distance-based-methods" class="nav-link" data-scroll-target="#distance-based-methods">Distance-based Methods</a>
  <ul class="collapse">
  <li><a href="#upgma" id="toc-upgma" class="nav-link" data-scroll-target="#upgma">UPGMA</a></li>
  <li><a href="#neighbour-joining" id="toc-neighbour-joining" class="nav-link" data-scroll-target="#neighbour-joining">Neighbour Joining</a></li>
  </ul></li>
  <li><a href="#character-based-methods" id="toc-character-based-methods" class="nav-link" data-scroll-target="#character-based-methods">Character-based Methods</a>
  <ul class="collapse">
  <li><a href="#parsimony" id="toc-parsimony" class="nav-link" data-scroll-target="#parsimony">Parsimony</a></li>
  <li><a href="#maximum-likelihood" id="toc-maximum-likelihood" class="nav-link" data-scroll-target="#maximum-likelihood">Maximum Likelihood</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Modeling Evolutionary Relationships</h1>
  <div class="quarto-categories">
    <div class="quarto-category">Phylogeny and Evolution</div>
    <div class="quarto-category">Alignment</div>
  </div>
  </div>

<div>
  <div class="description">
    Build trees to infer phylogenetic relationships across species.
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Jan Emmanuel Samson </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 28, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This article is inspired by a lecture series from Prof.&nbsp;Greg Kellog. Full playlist can be accessed <a href="https://www.youtube.com/playlist?list=PLjPAsbaa7pgGOrSPNBojU6lFCsJj7pvdJ">here</a>.</p>
</div>
</div>
<section id="evolution" class="level2">
<h2 class="anchored" data-anchor-id="evolution">Evolution</h2>
<section id="what-drives-change" class="level3">
<h3 class="anchored" data-anchor-id="what-drives-change">What drives change?</h3>
<p>Mutations in the genome create new alleles and change allele frequencies in a population, and as such drive the process of evolution. A simplified model of evolution has two rules:</p>
<ol type="1">
<li>Mutations that confer a fitness advantage are maintained and inherited across generations.</li>
<li>Mutations that deter fitness are selected against and purged from the population.</li>
</ol>
<p>This simplified model does not account for population dynamics and environmental factors that may possible affect the overall fitness of a population. From previous studies, we know that there are plenty of other factors that drive evolution such as:</p>
<ul>
<li>Negative selection</li>
<li>Positive selection</li>
<li>Genetic drift</li>
<li>Genetic expansion (duplication)</li>
<li>Genetic loss (deletions)</li>
<li>Horizontal gene transfer</li>
</ul>
<dl>
<dt>Negative Selection</dt>
<dd>
The natural selection <em>against</em> deleterious mutations, preserving functionally important sites.
</dd>
<dt>Postive Selection</dt>
<dd>
The process of Darwinian selection by which <em>new, advantageous</em> variants sweep a population.
</dd>
<dt>Genetic Drift</dt>
<dd>
Most mutations the occurred within a population are selectively <em>neutral</em> (Neutral Theory by Kimura).
</dd>
</dl>
</section>
<section id="how-do-we-observe-selection" class="level3">
<h3 class="anchored" data-anchor-id="how-do-we-observe-selection">How do we observe selection?</h3>
<p>Simplify the model by enforcing a few assumptions:</p>
<ul>
<li>Some changes are not subject to selection (neutral mutations)</li>
<li>Non-synonymous mutations in amino acids are subjected to greater selection pressure</li>
<li>The overall rate of synonymous mutation is a good approximation of neutral mutations</li>
</ul>
<p>Selection in protein-coding regions is then infered by:</p>
<p><span class="math display">\[
\frac{\text{non-synonymous mutation rate}}{\text{synonymous mutation rate}}
\]</span></p>
</section>
<section id="synonymous-vs.-non-synonymous-mutations" class="level3">
<h3 class="anchored" data-anchor-id="synonymous-vs.-non-synonymous-mutations">Synonymous vs.&nbsp;non-synonymous mutations</h3>
<ul>
<li>The degeneracy of th genetic code means some nucleotide mutations will not lead to protein change.</li>
<li>Transition/transversion occur at unequal rates</li>
<li>Positive selection protein sequences can be inferred from elevated rate of non-synonymous mutations over a neutral model</li>
</ul>
</section>
</section>
<section id="phylogenetics" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="phylogenetics">Phylogenetics</h2>
<p>The study of evolutionary relationships using genetic data, quantitative traits, characters, and morphological data matrices. The subfield of <strong>molecular phylogenetics</strong> focuses on the use of molecular data such as DNA and protein sequences to infer phylogenies. The end result of doing this type of analysis is usually a <strong>phylogenetic tree</strong> which is a branching diagram showing inferred evolutionary relationships between samples and species.</p>
<p>A phylogenetic tree is composed of:</p>
<ol type="1">
<li><strong>Internal nodes</strong> to represent ancestor species;</li>
<li><strong>External nodes</strong> (or leaves) to represent extant, known species; and</li>
<li><strong>Branches</strong> whose lengths represent the evolutionary distance between species</li>
</ol>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://www.yourgenome.org/wp-content/uploads/2023/11/1171-2022-10-03_phylogenetic-tree-diagram-yourgenome_CS_ES.png.webp" class="img-fluid figure-img"></p>
<figcaption>Components of a Phylogenetic Tree.</figcaption>
</figure>
</div>

<div class="no-row-height column-margin column-container"><div class="">
<p>Retrieved from <a href="https://www.yourgenome.org/theme/what-is-phylogenetics">Your Genome</a>.</p>
</div></div><section id="rooted-and-unrooted-trees" class="level3">
<h3 class="anchored" data-anchor-id="rooted-and-unrooted-trees">Rooted and unrooted trees</h3>
<ul>
<li>If a tree is rooted, the nodes repreent inferred <em>most recent common ancestor</em></li>
<li>Sometimes, the lengths of branches on a rooted tree can be meaningful
<ul>
<li>In a <strong>Phylogram</strong>, length represents the amount of character change</li>
<li>In a <strong>Chronogram</strong>, length represents evolutionary time</li>
</ul></li>
<li>unrooted trees may <em>look</em> rooted because of they way they are displayed</li>
</ul>
<p>A tree is deemed <strong>bifurcating</strong> if two descendants arise from every interior node in a phylogenetic tree, akin to the structure of binary trees. A more general term for diagrams that illustrate evolutionary relationships is the <strong>dendogram</strong>. All trees are dendograms, but not all dendograms are trees.</p>
</section>
<section id="how-do-we-generate-a-tree" class="level3">
<h3 class="anchored" data-anchor-id="how-do-we-generate-a-tree">How do we generate a tree?</h3>
<section id="decide-what-sequences-to-include-in-the-tree" class="level4">
<h4 class="anchored" data-anchor-id="decide-what-sequences-to-include-in-the-tree">1. Decide what sequences to include in the tree</h4>
<ul>
<li>This is the task of deciding the tips of the trees</li>
<li>It if often the most time-consuming</li>
<li>Think about the question your are trying to answer</li>
<li>Perform queries using BLAST</li>
</ul>
</section>
<section id="align-the-sequences-using-an-msa-tool." class="level4">
<h4 class="anchored" data-anchor-id="align-the-sequences-using-an-msa-tool.">2. Align the sequences using an MSA tool.</h4>
<ul>
<li>Use tools like CLUSTAL, MUSCLE, T-Coffee, etc.</li>
<li>There are scenarios where produced alignments need to be edited (e.g.&nbsp;end trimming) due to how tree inference algorithms work.</li>
</ul>
</section>
<section id="estimate-the-tree." class="level4">
<h4 class="anchored" data-anchor-id="estimate-the-tree.">3. Estimate the tree.</h4>
<dl>
<dt>Distance-based method</dt>
<dd>
After alignment, aligned sequences are converted into <em>distances</em> representing pairwise differences between sequences. A distance matix yields a single tree.
</dd>
<dt>Character-based method</dt>
<dd>
After alignment, each position in the alignment is compared. Homologous positions must be aligned. Multiple trees are generated and compared.
</dd>
</dl>
</section>
<section id="root-the-tree-if-possible." class="level4">
<h4 class="anchored" data-anchor-id="root-the-tree-if-possible.">4. Root the tree if possible.</h4>
<table class="caption-top table">
<colgroup>
<col style="width: 42%">
<col style="width: 57%">
</colgroup>
<thead>
<tr class="header">
<th>Method</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Outgroup</td>
<td>Place root between distant sequence and the rest of the tree</td>
</tr>
<tr class="even">
<td>Midpoint</td>
<td>Place root at midpoint of longest path between two leaves</td>
</tr>
<tr class="odd">
<td>Gene duplication</td>
<td>Place root bewteen paralogous gene copies</td>
</tr>
<tr class="even">
<td>UPGMA</td>
<td>Assume a molecular clocl</td>
</tr>
</tbody>
</table>
</section>
<section id="draw-the-tree-to-allow-interpretation." class="level4">
<h4 class="anchored" data-anchor-id="draw-the-tree-to-allow-interpretation.">5. Draw the tree to allow interpretation.</h4>
</section>
</section>
<section id="substitution-models" class="level3">
<h3 class="anchored" data-anchor-id="substitution-models">Substitution Models</h3>
<p>Attempts to estimate the evolutionary divergence from observed substitutions.</p>
<ul>
<li>Each such model forms a Markov process (similar to PAM), with two components:
<ul>
<li>A stationary distribution of nucleotide frequencies, <span class="math inline">\(\pi_A\)</span>, <span class="math inline">\(\pi_G\)</span>, <span class="math inline">\(\pi_C\)</span>, <span class="math inline">\(\pi_T\)</span>, <span class="math inline">\(\sum_\pi = 1\)</span></li>
<li>A transition rate matrix <span class="math inline">\(Q\)</span> that represent transtions that would occur by substitution</li>
</ul></li>
<li>A substitution model can be used to estimate evolutionary distance between two sequences.</li>
<li>A substitution model can also be used to simulate divergence.</li>
<li>Nucleotide substitution models range widely in complexity.</li>
<li>Phylogenetic tools may provide statistical tests (based on nucleotide sequence alignments and simulations) to help determine the most appropriate substitution model.</li>
</ul>
<section id="jukes-cantor-model" class="level4">
<h4 class="anchored" data-anchor-id="jukes-cantor-model"><strong>Jukes-Cantor Model</strong></h4>
<ul>
<li>All mutations are equal with an overall mutation rate <span class="math inline">\(α\)</span>.</li>
<li>Nucleotide frequencies are equally distributed, i.e., <span class="math inline">\(\pi_A\)</span> = <span class="math inline">\(\pi_G\)</span> = <span class="math inline">\(\pi_C\)</span> = <span class="math inline">\(\pi_T\)</span> = 0.25</li>
<li>Considers one parameters: <span class="math inline">\(α\)</span>.</li>
</ul>
<div class="columns center">
<div class="column" style="width:50%;">
<p><strong>Mutation probability matrix</strong></p>
<span class="math display">\[\begin{matrix}
  \quad &amp; A &amp; G &amp; C &amp; T \\
  A &amp; 1-3α &amp; α &amp; α &amp; α \\
  G &amp; α &amp; 1-3α &amp; α &amp; α \\
  C &amp; α &amp; α &amp; 1-3α &amp; α \\
  T &amp; α &amp; α &amp; α &amp; 1-3α \\
\end{matrix}\]</span>
</div><div class="column" style="width:50%;">
<p><strong>Transition rate matrix</strong></p>
<span class="math display">\[\begin{matrix}
  \quad &amp; A &amp; G &amp; C &amp; T \\
  A &amp; -3α &amp; α &amp; α &amp; α \\
  G &amp; α &amp; -3α &amp; α &amp; α \\
  C &amp; α &amp; α &amp; -3α &amp; α \\
  T &amp; α &amp; α &amp; α &amp; -3α \\
\end{matrix}\]</span>
</div>
</div>
<p>Evolutionary distance can be computed as:</p>
<p><span class="math display">\[
D_{xy} = -\frac{3}{4} ln(1-\frac{4}{3}p)
\]</span></p>
<p>where <span class="math inline">\(D_{xy}\)</span> is the distance between sequences <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> in terms of number of changes per site, an <span class="math inline">\(p\)</span> is the <em>observed</em> proportion of sites that differ between them.</p>
</section>
<section id="kimura-two-parameter-model" class="level4">
<h4 class="anchored" data-anchor-id="kimura-two-parameter-model"><strong>Kimura Two-Parameter Model</strong></h4>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Assigned different mutation rates for transitions and transversions
<ul>
<li><span class="math inline">\(α\)</span> for transitions</li>
<li><span class="math inline">\(β\)</span> for transversions</li>
</ul></li>
<li>all base frequencies are equal
<ul>
<li>i.e., <span class="math inline">\(\pi_A\)</span> = <span class="math inline">\(\pi_G\)</span> = <span class="math inline">\(\pi_C\)</span> = <span class="math inline">\(\pi_T\)</span> = 0.25</li>
</ul></li>
</ul>
</div><div class="column center" style="width:50%;">
<p><strong>K2P Transition rate matrix</strong></p>
<span class="math display">\[\begin{matrix}
  \quad &amp; A &amp; G &amp; C &amp; T \\
  A &amp; -α-2β &amp; α &amp; α &amp; α \\
  G &amp; α &amp; -α-2β &amp; α &amp; α \\
  C &amp; α &amp; α &amp; -α-2β &amp; α \\
  T &amp; α &amp; α &amp; α &amp; -α-2β \\
\end{matrix}\]</span>
</div>
</div>
</section>
<section id="falsenstein-tajima-nei-model" class="level4">
<h4 class="anchored" data-anchor-id="falsenstein-tajima-nei-model"><strong>Falsenstein-Tajima-Nei Model</strong></h4>
<ul>
<li>Instead of varying transition/transversion rates, Falsenstein allowed varying nucleotide frequences:</li>
</ul>
<p><span class="math display">\[π_A \ne π_G \ne π_C \ne π_T, \sum_{\pi} = 1\]</span></p>
<ul>
<li>The relative probability of mutating to a particular base <span class="math inline">\(i\)</span> was given by its frequence <span class="math inline">\(\pi_i\)</span>.</li>
<li>Introduced a maximum likelihood method for inferring phylogenies.</li>
</ul>
</section>
<section id="hasegawa-kishino-and-yano-1985" class="level4">
<h4 class="anchored" data-anchor-id="hasegawa-kishino-and-yano-1985"><strong>Hasegawa, Kishino, and Yano 1985</strong></h4>
<ul>
<li>Combined K2P (<span class="math inline">\(α\)</span> and <span class="math inline">\(β\)</span>) with Felsenstein (varying <span class="math inline">\(π\)</span>)</li>
</ul>
</section>
<section id="other-models" class="level4">
<h4 class="anchored" data-anchor-id="other-models"><strong>Other Models</strong></h4>
<ul>
<li>More complicated models up to allowing all substitution frequencies and nucleotide frequencies (i.e., Generalized Time-Reversible) can be derived (Tavare et al.&nbsp;1986)</li>
<li>It is also possible to have different base rates at different positions (which might be expected in coding sequnces) with a different estimator <span class="math inline">\(γ\)</span>.</li>
<li>More complicated models have more parameters.</li>
<li>The more complicated models do not necessarily perform better. The most appropriate model may be assessed by simulation of different models and statistical tests to compare them.</li>
<li>Goal is to choose the model that is most appropriate for the data.</li>
</ul>
</section>
</section>
</section>
<section id="distance-based-methods" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="distance-based-methods">Distance-based Methods</h2>
<p>When aligning sequence <span class="math inline">\(S_i\)</span> with sequence <span class="math inline">\(S_j\)</span>,</p>
<ul>
<li>Calculate a <em>dissimilarity</em> matrix that has some measure of <em>distance</em> between sequences. We call this <span class="math inline">\(D_{ij}\)</span>.</li>
<li>For <span class="math inline">\(D_{ij}\)</span> in guide tree construction, we often use the % mismatch between <span class="math inline">\(S_i\)</span> and <span class="math inline">\(S_j\)</span> at aligned positions.</li>
<li>For <span class="math inline">\(D_{ij}\)</span> in phylogenetics, we may estimate a distance from the observed differences and a substitution model.</li>
</ul>
<p>Metric distances always object the triangle inequality formula:</p>
<p><span class="math display">\[ D_{ik} \le D_{ij} + D_{jk} \]</span></p>
<p>A uniform molecular clock assumes an <em>ultrametic</em> distance:</p>
<p><span class="math display">\[ D_{ik} \le \text{max}(D_{ij}, D_{jk}) \]</span></p>
<section id="upgma" class="level3">
<h3 class="anchored" data-anchor-id="upgma">UPGMA</h3>
<p><strong>Unweighted Pair Group Mean Average</strong></p>
<ol type="1">
<li>The pair (row and column) with the smallest distance in the matrix is joined.</li>
<li>The pair becomes a cluster, with an interior node positioned halfway between the entries.</li>
<li>A new distance matrix is calculated with the cluster replacing the individual entries using an average of all contributing pairs</li>
<li>Repeat until done.</li>
</ol>
<p>The algorithm runs at <span class="math inline">\(O(n^2 \log n)\)</span> time complexity and produces a unique, rooted tree.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Intro</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">1a</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-3" role="tab" aria-controls="tabset-1-3" aria-selected="false">1b</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-4-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-4" role="tab" aria-controls="tabset-1-4" aria-selected="false">2a</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-5-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-5" role="tab" aria-controls="tabset-1-5" aria-selected="false">2b</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-6-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-6" role="tab" aria-controls="tabset-1-6" aria-selected="false">2c</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-7-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-7" role="tab" aria-controls="tabset-1-7" aria-selected="false">3a</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-8-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-8" role="tab" aria-controls="tabset-1-8" aria-selected="false">3b</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-9-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-9" role="tab" aria-controls="tabset-1-9" aria-selected="false">3c</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-10-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-10" role="tab" aria-controls="tabset-1-10" aria-selected="false">4a</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-11-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-11" role="tab" aria-controls="tabset-1-11" aria-selected="false">4b</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-12-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-12" role="tab" aria-controls="tabset-1-12" aria-selected="false">5</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-13-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-13" role="tab" aria-controls="tabset-1-13" aria-selected="false">6</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-14-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-14" role="tab" aria-controls="tabset-1-14" aria-selected="false">End</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma01.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma02.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-3-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma03.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-4" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-4-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma04.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-5" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-5-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma05.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-6" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-6-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma06.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-7" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-7-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma07.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-8" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-8-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma08.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-9" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-9-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma09.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-10" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-10-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma10.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-11" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-11-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma11.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-12" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-12-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma12.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-13" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-13-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma13.PNG" class="img-fluid"></p>
</div>
<div id="tabset-1-14" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-14-tab">
<p><img src="http://www.slimsuite.unsw.edu.au/teaching/upgma/upgma14.PNG" class="img-fluid"></p>
</div>
</div>
</div>
</section>
<section id="neighbour-joining" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="neighbour-joining">Neighbour Joining</h3>
<p>Neighbour joining is guaranteed to yield a unique tree from a set of additive distances. The model does not assume equal rates of evolution on all lineages and thus leads to an unrooted tree. Its goal is to minimize the total branch length of a tree using an objective function.</p>
<p>The steps of the algorithm are as follows:</p>
<ol type="1">
<li>Start with a distance matrix <span class="math inline">\(D\)</span> among sequences.</li>
<li>Calculate the net divergence <span class="math inline">\(Q_{ij}\)</span> of each taxon pair among the <span class="math inline">\(n\)</span> sequences:</li>
</ol>

<div class="no-row-height column-margin column-container"><div class="">
<p><span class="math display">\[R_i = \sum_{k=1}^{n} D_{ik}\]</span></p>
<p>where <span class="math inline">\(R_i\)</span> is the total distance from a sequence <span class="math inline">\(S_i\)</span> to all others.</p>
</div></div><p><span class="math display">\[ Q_{ij} = (n-2)D_{ij} - R_i - R_j \]</span></p>
<ol start="3" type="1">
<li>Compute the distances between the two points (<span class="math inline">\(f\)</span> and <span class="math inline">\(g\)</span>) and a new node labeled as <span class="math inline">\(u\)</span>.</li>
</ol>
<p><span class="math display">\[ δ(f,u) = \frac{1}{2} D_{fg} + \frac{1}{2(n-2)} [R_f - R_g] \]</span></p>
<p><span class="math display">\[ δ(g,u) = δ(f,g) - δ(f,u) \]</span></p>
<ol start="4" type="1">
<li>Compute the distance between <span class="math inline">\(u\)</span> and the remaining nodes <span class="math inline">\(k\)</span>.</li>
</ol>
<p><span class="math display">\[ D_{uk} = \frac{1}{2}[D_{fk} + D_{gk} - D_{fg}] \]</span></p>
<ol start="5" type="1">
<li>Recompute the distance matrix and repeat until done.</li>
</ol>
</section>
</section>
<section id="character-based-methods" class="level2">
<h2 class="anchored" data-anchor-id="character-based-methods">Character-based Methods</h2>
<p>Looks at each column in the alignment.</p>
<section id="parsimony" class="level3">
<h3 class="anchored" data-anchor-id="parsimony">Parsimony</h3>
<ul>
<li>Seeks the minimum number of changes and often results in several equivalent trees.</li>
<li>Found in a variety of computer science algorithms:
<ul>
<li>heuristic methods</li>
<li>branch and bound</li>
<li>nearest neighbor interchange</li>
</ul></li>
</ul>
<section id="what-happens-when-a-character-evolves-rapidly" class="level4">
<h4 class="anchored" data-anchor-id="what-happens-when-a-character-evolves-rapidly">What happens when a character evolves rapidly?</h4>
<ul>
<li>in DNA, a rapidly evolving site might be inferred to be unchanged, when in reality it will have changed twice
<ul>
<li>A → C → A</li>
</ul></li>
<li>This can lead to statistically inconsistent results</li>
<li>Maximum Parsimony is particularly prone to this behavior</li>
<li>The more data, the more likely to be wrong</li>
<li>Various detection and solution methods, all imperfect</li>
</ul>
</section>
</section>
<section id="maximum-likelihood" class="level3">
<h3 class="anchored" data-anchor-id="maximum-likelihood">Maximum Likelihood</h3>
<ul>
<li>Under a model of evolution, generates a tree that maximizes the likelihood of observing the data.</li>
<li>Produces different trees with varying likelihood (measured as <span class="math inline">\(L\)</span>).</li>
</ul>


</section>
</section>

<a onclick="window.scrollTo(0, 0); return false;" role="button" id="quarto-back-to-top"><i class="bi bi-arrow-up"></i> Back to top</a></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      const annoteTargets = window.document.querySelectorAll('.code-annotation-anchor');
      for (let i=0; i<annoteTargets.length; i++) {
        const annoteTarget = annoteTargets[i];
        const targetCell = annoteTarget.getAttribute("data-target-cell");
        const targetAnnotation = annoteTarget.getAttribute("data-target-annotation");
        const contentFn = () => {
          const content = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          if (content) {
            const tipContent = content.cloneNode(true);
            tipContent.classList.add("code-annotation-tip-content");
            return tipContent.outerHTML;
          }
        }
        const config = {
          allowHTML: true,
          content: contentFn,
          onShow: (instance) => {
            selectCodeLines(instance.reference);
            instance.reference.classList.add('code-annotation-active');
            window.tippy.hideAll();
          },
          onHide: (instance) => {
            unselectCodeLines();
            instance.reference.classList.remove('code-annotation-active');
          },
          maxWidth: 300,
          delay: [50, 0],
          duration: [200, 0],
          offset: [5, 10],
          arrow: true,
          appendTo: function(el) {
            return el.parentElement.parentElement.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto',
          placement: 'right',
          popperOptions: {
            modifiers: [
            {
              name: 'flip',
              options: {
                flipVariations: false, // true by default
                allowedAutoPlacements: ['right'],
                fallbackPlacements: ['right', 'top', 'top-start', 'top-end', 'bottom', 'bottom-start', 'bottom-end', 'left'],
              },
            },
            {
              name: 'preventOverflow',
              options: {
                mainAxis: false,
                altAxis: false
              }
            }
            ]        
          }      
        };
        window.tippy(annoteTarget, config); 
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
<p>© 2024, Jan Samson</p>
</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>